llvm/lib/LTO/ThinLTOCodeGenerator.cpp

```diff
@@ -439,6 +445,233 @@ class ModuleCacheEntry {
+    return ModuleCacheEntry::writeObject(OutputBuffer, OutputPath);
+  }
+
+  Optional<std::unique_ptr<MemoryBuffer>> getMappedBuffer() final {
+    if (getEntryPath().empty())
+      return None;
+
+    auto ReloadedBufferOrErr = tryLoadingBuffer();
+    if (auto EC = ReloadedBufferOrErr.getError()) {
+      // On error, keep the preexisting buffer and print a diagnostic.
+      errs() << "remark: can't reload cached file '" << getEntryPath()
+             << "': " << EC.message() << "\n";
+    }
+    return std::move(*ReloadedBufferOrErr);
+  }
+
+private:
+  SmallString<128> EntryPath;
+};
+
+class CASModuleCacheEntry : public ModuleCacheEntry {
+public:
+  // Create a cache entry. This compute a unique hash for the Module considering
+  // the current list of export/import, and offer an interface to query to
+  // access the content in the cache.
+  CASModuleCacheEntry(
+      cas::ObjectStore &CAS, cas::ActionCache &Cache,
+      const ModuleSummaryIndex &Index, StringRef ModuleID,
+      const FunctionImporter::ImportMapTy &ImportList,
+      const FunctionImporter::ExportSetTy &ExportList,
+      const std::map<GlobalValue::GUID, GlobalValue::LinkageTypes> &ResolvedODR,
+      const GVSummaryMapTy &DefinedGVSummaries, unsigned OptLevel,
+      bool Freestanding, const TargetMachineBuilder &TMBuilder)
+      : CAS(CAS), Cache(Cache) {
+    Optional<std::string> Key =
+        computeCacheKey(Index, ModuleID, ImportList, ExportList, ResolvedODR,
+                        DefinedGVSummaries, OptLevel, Freestanding, TMBuilder);
+
+    if (!Key)
+      return;
+
+    // Create the key by inserting cache key (SHA1) into CAS to create a ID for
+    // the correct context.
+    // TODO: We can have an alternative hashing function that doesn't
+    // need to store the key into CAS to get the CacheKey.
+    auto CASKey = CAS.createProxy(None, *Key);
+    if (!CASKey)
+      report_fatal_error(CASKey.takeError());
+
+    ID = CASKey->getID();
+  }
+
+  std::string getEntryPath() final {
+    if (!ID)
+      return "";
+
+    return ID->toString();
+  }
+
+  // Try loading the buffer for this cache entry.
+  ErrorOr<std::unique_ptr<MemoryBuffer>> tryLoadingBuffer() final {
+    if (!ID)
+      return std::error_code();
+
+    auto MaybeKeyID = Cache.get(*ID);
+    if (!MaybeKeyID)
+      return errorToErrorCode(MaybeKeyID.takeError());
+
+    if (!*MaybeKeyID)
+      return std::error_code();
+
+    auto MaybeObject = CAS.getProxy(**MaybeKeyID);
+    if (!MaybeObject)
+      return errorToErrorCode(MaybeObject.takeError());
+
+    return MaybeObject->getMemoryBuffer("", /*NullTerminated=*/true);
+  }
+
+  // Cache the Produced object file
+  void write(const MemoryBuffer &OutputBuffer) final {
+    if (!ID)
+      return;
+
+    auto Proxy = CAS.createProxy(None, OutputBuffer.getBuffer());
+    if (!Proxy)
+      report_fatal_error(Proxy.takeError());
+
+    auto Err = Cache.put(*ID, Proxy->getID());
+    report_fatal_error(std::move(Err));
```

Comment:
Need to check for error before calling `report_fatal_error`, right? 🙂

Reply:
Nice catch! Fixed.